## Wendy's Diet Problem
## OPIM 5641 BDM Project 1 
##Team 3

Vedant Dadhich

Lauren Ein

	
Junchi Ma

Jiajun Zhou

## Set up the environment 

We will import the necessary libraries..

In [ ]:
# before you do anything...
# mount your drive!
# click folder on the left...

%matplotlib inline
from pylab import *

import shutil
import sys
import os.path

if not shutil.which("pyomo"):
    !pip install -q pyomo
    assert(shutil.which("pyomo"))

if not (shutil.which("glpsol") or os.path.isfile("glpsol")):
    if "google.colab" in sys.modules:
        !apt-get install -y -qq glpk-utils
    else:
        try:
            !conda install -c conda-forge ipopt
        except:
            pass

assert(shutil.which("glpsol") or os.path.isfile("glpsol"))

from pyomo.environ import *

SOLVER = 'glpk'
EXECUTABLE = '/usr/bin/glpsol'

     |████████████████████████████████| 9.4MB 580kB/s 
     |████████████████████████████████| 256kB 33.4MB/s 
     |████████████████████████████████| 51kB 5.2MB/s 
     |████████████████████████████████| 163kB 35.7MB/s 
Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 144611 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-

## Describe the diet problem


### 1. Introduction 

 The FDA also known as the United States Food and Drug Administration is one of the government’s federal agencies of the Department of Health and Human Services. The agency is responsible for protecting public health and safety through regulation of food, human and veterinary drugs, vaccines, and other biological products. 

In terms of food, in addition to the focus on food safety, the FDA has put out guidance on food/nutrition labeling. FDA does not pre-approve food labels but rather they are assuring to food sold in the US whether they are domestically or foreign produced are safe, wholesome, and properly label. 
The article issued out by the FDA on May 5, 2020 called “Daily Value on the New Nutrition and Supplement Facts Labels”< https://www.fda.gov/food/new-nutrition-facts-label/daily-value-new-nutrition-and-supplement-facts-labels#referenceguide> is the source of the nutrition label and daily values used in our model. We choose this diet as the base of the constraints of our model because the labeling of the food we buy and eat are based on guidelines from the FDA and their goal is to ensure the health and safety of the food in America.

Looking at the Nutrition Facts label itself, the list must include total fat, saturated fat, trans fat, cholesterol, sodium, total carbohydrate, dietary fiber, total sugars, added sugars, protein, and in some cases certain vitamins and minerals. On the label it is further broken down into daily values. Given the Wendy’s menu data the areas we will focus on satisfying the nutritional requirements in the areas: Calories, Protein, Total Carbs, Sodium and Sugars.

Wendy's is an American fast food chain restaurant founded in 1969, in Columbus, Ohio.  Wendy's was the world's third-largest hamburger fast-food chain with 6,711 location worldwide. The chain is known for its square hamburgers, sea salt fries, and Frosty, a form of soft-serve ice cream mixed with starches. The Wendy's menu data set the model is based off is 145 rows and split into categories hamburgers, chicken sandwiches, sides, entrees, Frosty, and beverages. 
 
Additionally the objective of the model is to minimize Weight Watcher points. For context the point system is supposed to be an easy to use tool help people control or lose weight by influencing a person’s lifestyle and eating habits. The Weight Watcher points were given in the dataset for each Wendy’s menu item, the estimated point value of each food is determined by a certain set of known parameters based on calories, sugar, saturated fat, and protein.

<p align="center">
<img src="https://www.fda.gov/files/NewNFL_5Group.jpg">
</p>

### 2. Declare the decision variable

We have 145 kinds of Wendy's items with their five types of nutrution values and the value of each weight watcher in this dataset [145 rows x 7 columns]. In this diet problem, we need decide what Wendy's items of products and the number of each item to cosume everyday in order to select one of the best combinations of products in Wendy's Menu satisfying nutritional requirements while minimizing Weight Watcher points.

Let's suppose the decison variables are $[X_1,X_2,...,X_{145}]$



### 3. Restrict the constraints



* Calories: Calories are a unit of energy or heat, in nutrition is actually 1,000 of these small calories. American food labels are reported in kilocalories, or kilojoules. For example, one medium-size apple contains 95 calories or it actually contains 95 kilocalories. This variable was used because the FDA food labels and the diet is based on a maximum of 2,000 Calories a day. <https://www.livescience.com/52802-what-is-a-calorie.html> 
  * Calorie constraint defined as: $\Sigma$ ( the number of each item $(X_i)$ * its calorie value ) $\geq$ 2000

* Protein(g): Proteins are in every cell of the human body. When protein foods are broken down into parts called amino acids during digestion and the human body various amino acids in large quantities to maintain good health. Amino acids are found in animal sources such as meats, milk, fish, and eggs. As well as in plant sources such as soy, beans, legumes, nut butters, and some grains. Lots of the Wendy’s food menu fall into one of these categories like their burgers and chicken sandwhiches; as stated proteins/amino acids are needed to maintain good health <https://medlineplus.gov/ency/article/002467.htm> 
  * Protein constraint defined as: $\Sigma$ ( the number of each item $(X_i)$ * its protein value ) $\geq$ 50

* Total Carbs (g):Carbohydrates are a main type of nutrients. Your digestive system changes carbohydrates into glucose and it used for energy for your cells, tissues and organs. There are two types of carbohydrates: simple and complex. Simple carbohydrates include natural and added sugars. Complex carbohydrates include whole grain breads and cereals, starchy vegetables and legumes. Similarly the energy is needed by your body and it is a component to constraint given in the Wendy’s menu items like the bun on the burger and sandwhiches<https://medlineplus.gov/definitions/nutritiondefinitions.html> 
  * Total Carbohydrates defined as: $\Sigma$ ( the number of each item $(X_i)$ * its value of total carbohydrates ) $\geq$ 275

* Sodium (g): Sodium or commonly known as table salt is made up of sodium and chlorine. Your body needs some sodium to work properly as it with the function of your nerves and muscles. It also helps to keep the right balance of fluids in your body. This is also a key part of a balance healthy diet. And used on stple items on the Wendy's menu for example the french fries <https://medlineplus.gov/definitions/nutritiondefinitions.html>
  * Sodium constraint (mg) defines as $\Sigma$ ( the number of each item $(X_i)$ * its sodium value ) $\geq$ 2300

* Sugar (g): Sugar is a type of carbohydrate and have a sweet taste. They can be found naturally in fruits, vegetables, milk, and milk products. There are three types of sugars: glucose, fructose, and sucrose They are also added to many foods and drinks during preparation or processing. Wendy’s is considered a fast food restaurant which serves Frosty’s, soda and other sugary drinks as well as the bun bread on a burger will also contain sugar. Therefore the dietary limitations on sugar is a needed constraint. <https://medlineplus.gov/definitions/nutritiondefinitions.html> 
  * Sugars constraint (g) defined as $\Sigma$ ( the number of each item $(X_i)$ * its sugars value ) $\geq$ 50




### 4. Define the objective function & model

Since the objective of the model is to minimize Weight Watcher points, we define the obejective function as MIN (Weight Watcher Points)= $\Sigma$(the number of each item $(X_i)$ *its value of weight watcher)

## Build the first model

### 1. Import the file 

Data is gathered from the https://www.nutrition-charts.com/wendys-nutrition-facts-calorie-information/ and we copy/paste the data into Excel with proper formatting. We use following code to upload the excel files and the data has 145 items and 6 variables 

In [ ]:
# Read data from Excel file
from google.colab import files
# Command to upload file
uploaded = files.upload()

Saving Project 1.xlsx to Project 1.xlsx


In [ ]:
import pandas as pd
import io

# Transform excel file in a pandas dataframe
df = pd.read_excel('Project 1.xlsx') 
df.head()

Item Name  Calories  ...  Sugars (g)  Weight Watchers
0            Apple Pecan Chicken Salad       350  ...          17                9
1  Apple Pecan Chicken Salad Half-Size       180  ...           9                5
2                       Roasted Pecans       110  ...           4                3
3     Pomegranate Vinaigrette Dressing        60  ...           8                2
4                     Baja Chili Salad       530  ...          10               15

[5 rows x 7 columns]

### 3. Extract all the needed values from the data

We extract the items of products and their names, the nutrients of each product and also their weight watcher points

In [ ]:
# List of product indices 
products = df.index.values.tolist()
print("products",products)

# Extract list of item names, use transpose operation, and create dictionary index->food name
products_name =  df["Item Name"].T.to_dict()
print("products_name",products_name)

# List of nutrients
nutrients = list(df)[1:6]
print("nutrients",nutrients)

# Extract nutritional values;
nut_value = dict()
for nutrient in nutrients:
  nut_value[nutrient] = df[nutrient].tolist()

print(nut_value)

# Extract weight watchers with values, transform column in list, and remove last row 
weight_watchers = df["Weight Watchers"].tolist()
print("weight watchers",weight_watchers)

products [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144]
products_name {0: 'Apple Pecan Chicken Salad', 1: 'Apple Pecan Chicken Salad Half-Size', 2: 'Roasted Pecans', 3: 'Pomegranate Vinaigrette Dressing', 4: 'Baja Chili Salad', 5: 'Baja Chili Salad Half-Size', 6: 'Seasoned Tortilla Strips', 7: 'Red Jalapeño Dressing', 8: 'Chicken BLT Cobb Salad', 9: 'Chicken BLT Cobb Salad Half-Size', 10: 'Avocado Ranch Dressing', 11: 'Spicy Ch

### 4. Set up the model of all items 

According to our research, the RHS values for constrraints calories, protein, total carb, sodium and sugar is 2000cal, 50g, 275g, 2300mg,and 50g. We set up the decision variables, constrains and objective function as we defined previously.

In [ ]:
# declare the model
model = ConcreteModel()

# declare decision variables
model.volume = Var(products,domain=NonNegativeReals)

# declare objective
obj_expr = 0
for product in products:
  obj_expr += weight_watchers[product]*model.volume[product]
model.weight_watchers = Objective(expr = obj_expr, sense = minimize)

# declare five constraints
model.Calories = ConstraintList()
Calories_exp = 0
for product in products:
  Calories_exp += nut_value["Calories"][product]*model.volume[product]
model.Calories.add(expr = Calories_exp >= 2000)


model.Protein = ConstraintList()
Protein_exp = 0
for product in products:
  Protein_exp += nut_value["Protein (g)"][product]*model.volume[product]
model.Protein.add(expr = Protein_exp >= 50)

model.Totalcarb = ConstraintList()
Totalcarb_exp = 0
for product in products:
  Totalcarb_exp += nut_value["Total Carbs (g)"][product]*model.volume[product]
model.Totalcarb.add(expr = Totalcarb_exp >= 275)

model.Sodium = ConstraintList()
Sodium_exp = 0
for product in products:
  Sodium_exp += nut_value["Sodium (mg)"][product]*model.volume[product]
model.Sodium.add(expr = Sodium_exp >= 2300)

model.Sugars = ConstraintList()
Sugars_exp = 0
for product in products:
  Sugars_exp += nut_value["Sugars (g)"][product]*model.volume[product]
model.Sugars.add(expr = Sugars_exp >= 50)

# show the model
model.pprint()

6 Set Declarations
    Calories_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    1 :    {1,}
    Protein_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    1 :    {1,}
    Sodium_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    1 :    {1,}
    Sugars_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    1 :    {1,}
    Totalcarb_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    1 :    {1,}
    volume_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :  145 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21

### 5. Solve the model and get the solution

In [ ]:
# solve it
SolverFactory(SOLVER, executable=EXECUTABLE).solve(model).write()

solution = pd.DataFrame.from_dict(model.volume.extract_values(), orient='index', columns=[str(model.volume)])


# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 11.3636363636364
  Upper bound: 11.3636363636364
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 141
  Number of nonzeros: 636
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.013315200805664062
# ----------------------------------------------------------
#   Solution Information
# ----------

In [ ]:
df['volume'] = solution['volume']
df[df['volume']>0]

Item Name  ...      volume
8                   Chicken BLT Cobb Salad  ...    1.136364
39                       Ketchup, 1 Packet  ...   22.727273
46  Minute Maid® Light Lemonade, Small Cup  ...  268.181818

[3 rows x 8 columns]

In [ ]:
print("Weight watchers = ", model.weight_watchers())

sol = df[df['volume']>0]

print("Calories for the selcted items = ", sum(sol['Calories']*sol['volume']),'cal')
print("Protein for the selcted items = ", sum(sol['Protein (g)']*sol['volume']),'g')
print("Total Carbs for the selcted items = ", sum(sol['Total Carbs (g)']*sol['volume']),'g')
print("Sodium for the selcted items = ", sum(sol['Sodium (mg)']*sol['volume']),'mg')
print("Sugars for the selcted items = ", sum(sol['Sugars (g)']*sol['volume']),'g')

Calories for the selcted items =  2000.0000000000005 cal
Protein for the selcted items =  50.00000000000016 g
Total Carbs for the selcted items =  326.13636363636346 g
Sodium for the selcted items =  4284.090909090911 mg
Sugars for the selcted items =  49.999999999999964 g


The first solution for weight watchers point is 11.36 and the number of items selected (obejctive function coefficients) are:
* Chicken BLT Cobb Salad 1.13636363636364
* Ketchup, 1 Packet 22.7272727272727
* Minute Maid® Light Lemonade, Small Cup 268.181818181818

The first model is quite disgusting and unrealistic because we can't consume 268 cups of lemonade everyday.

### Report sensitivity analysis
 

In [ ]:
# First, we will save the model (you will see the file model.lp showing up on the left after executing the line below)
model.write("/content/model.lp", io_options={'symbolic_solver_labels': True})

# After running the line below, we will generate the file "sensit.sen", which contains the report we want to see
!/usr/bin/glpsol -m /content/model.lp --lp --ranges sensit.sen

# After generating the report, we want to see its content:
!cat /content/sensit.sen

GLPSOL: GLPK LP/MIP Solver, v4.65
Parameter(s) specified in the command line:
 -m /content/model.lp --lp --ranges sensit.sen
Reading problem data from '/content/model.lp'...
6 rows, 141 columns, 636 non-zeros
935 lines were read
GLPK Simplex Optimizer, v4.65
6 rows, 141 columns, 636 non-zeros
Preprocessing...
5 rows, 140 columns, 635 non-zeros
Scaling...
 A: min|aij| =  1.000e+00  max|aij| =  1.990e+03  ratio =  1.990e+03
GM: min|aij| =  1.267e-01  max|aij| =  7.892e+00  ratio =  6.229e+01
EQ: min|aij| =  1.605e-02  max|aij| =  1.000e+00  ratio =  6.229e+01
Constructing initial basis...
Size of triangular part is 5
      0: obj =   0.000000000e+00 inf =   1.263e+01 (5)
      5: obj =   6.038461538e+01 inf =   0.000e+00 (0)
*    11: obj =   1.136363636e+01 inf =   0.000e+00 (0)
OPTIMAL LP SOLUTION FOUND
Time used:   0.0 secs
Memory used: 0.2 Mb (185540 bytes)
Write sensitivity analysis report to 'sensit.sen'...
GLPK 4.65 - SENSITIVITY ANALYSIS REPORT                                     

From the sensitivity analysis report, we can see that:
* LHS value:<br>
Calories for the selcted items =  2000 cal<br>
Protein for the selcted items =  50 g<br>
Total Carbs for the selcted items =  326 g<br>
Sodium for the selcted items =  4284 mg<br>
Sugars for the selcted items =  50 g<br>

* RHS value:<br>
Calories for the selcted items =  2000 cal<br>
Protein for the selcted items =  50 g<br>
Total Carbs for the selcted items =  275 g<br>
Sodium for the selcted items =  2300 mg<br>
Sugars for the selcted items =  50 g<br>

* the Constriants Calorie , sugar and protein are binding, because the LHS and RHS of these three constraints are equal. <br> The reason why binding constraints are important is that these binding constraints give us limits to stop us from getting the better solution, so binding constraints are the one we need working on and find better way to distribute.




